openebs/cmd/openebs-provisioner/main.go

```
@@ -0,0 +1,113 @@
+	} else {
+		// Create an InClusterConfig and use it to create a client for the controller
+		// to use to communicate with Kubernetes
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		glog.Errorf("Failed to create config: %v", err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Errorf("Failed to create client: %v", err)
+	}
+
+	// The controller needs to know what the server version is because out-of-tree
+	// provisioners aren't officially supported until 1.5
+	serverVersion, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		glog.Errorf("Error getting server version: %v", err)
+	}
+
+	// Create the provisioner: it implements the Provisioner interface expected by
+	// the controller
+	var openEBSProvisioner controller.Provisioner
+	if CASTemplateFeatureGate() {
+		glog.Infof("Using CAS template feature gate for volume provisioning")
+		openEBSProvisioner = vol.NewOpenEBSProvisionerV1alpha1(clientset)
+	} else {
+		openEBSProvisioner = vol.NewOpenEBSProvisioner(clientset)
+	}
+	if openEBSProvisioner != nil {
+		// Start the provision controller which will dynamically provision OpenEBS VSM
+		// PVs
+		pc := controller.NewProvisionController(
+			clientset,
+			provisionerName,
+			openEBSProvisioner,
+			serverVersion.GitVersion)
+
+		pc.Run(wait.NeverStop)
+	} else {
+		os.Exit(1) //Exit if provisioner not created.
+	}
+
+}
+
+const (
+	// CASTemplateFeatureGateENVK is the ENV key to fetch cas template feature gate
+	// i.e. if cas template based provisioning is enabled or disabled
+	CASTemplateFeatureGateENVK = "OPENEBS_IO_CAS_TEMPLATE_FEATURE_GATE"
+)
+
+// CASTemplateFeatureGate returns true if cas template feature gate is
+// enabled
+func CASTemplateFeatureGate() bool {
+	// getEnv fetches the environment variable value from the runtime's environment
+	feature, _ := strconv.ParseBool(os.Getenv(string(CASTemplateFeatureGateENVK)))
```

Let us error out. Lets not do a underscore. Silent errors are dangerous.

---

Yes , next subsequent commits will handle some of these errors.